### PR TITLE
[Experimental] Allow animations to be started from a background thread

### PR DESCRIPTION
**Pull request by kmadsen**
This pull request is for communication and investigation purposes. I'm also running into enough issues that, I want to write them down somewhere.

## Issues 

https://github.com/mapbox/mapbox-maps-android/issues/1413
https://github.com/mapbox/mapbox-maps-android/issues/1556

In order to make Android Auto animations work with Xiaomi devices, all animations need to be run on a background thread. This is because the Xiaomi devices ignore the [Choreographer](https://developer.android.com/reference/android/view/Choreographer) calls when the app is in background. The CameraAnimator uses ValueAnimator, and ValueAnimator uses Choreographer.

## More issues

Mapbox has many calls that assume the UI thread, so in order to trigger the animations from a background thread there would be a bit work to ensure each call receives appropriate threads.

Here is an example where an evaluator is running on a background thread and Mapbox Maps assumes ui thread
```
2022-08-16 13:56:33.589 14515-14912/com.mapbox.maps.testapp.auto I/Mapbox: [maps-android\kyle_debug]: getFlyTo.CameraCenterAnimator.evaluator animation_thread_c8b20b45-befb-499f-8a4e-f798c442284b
2022-08-16 13:56:33.590 14515-14912/com.mapbox.maps.testapp.auto E/AndroidRuntime: FATAL EXCEPTION: animation_thread_c8b20b45-befb-499f-8a4e-f798c442284b
    Process: com.mapbox.maps.testapp.auto, PID: 14515
    com.mapbox.maps.exception.WorkerThreadException: The exception that is thrown when an application attempts to 
    perform a map operation on a worker thread.
        at com.mapbox.maps.ThreadChecker.throwIfNotMainThread(ThreadChecker.kt:88)
        at com.mapbox.maps.MapboxMap.checkNativeMap(MapboxMap.kt:1792)
        at com.mapbox.maps.MapboxMap.checkNativeMap$default(MapboxMap.kt:1790)
        at com.mapbox.maps.MapboxMap.unproject(MapboxMap.kt:950)
        at com.mapbox.maps.plugin.animation.CameraAnimatorsFactory.getFlyTo$lambda-9(CameraAnimatorsFactory.kt:377)
        at com.mapbox.maps.plugin.animation.CameraAnimatorsFactory.$r8$lambda$i0eVRz5Pdg1wufkPMzLCax34Kqw(Unknown Source:0)
        at com.mapbox.maps.plugin.animation.CameraAnimatorsFactory$$ExternalSyntheticLambda1.evaluate(Unknown Source:41)
```

### Before - animations break when app is backgrounded

https://user-images.githubusercontent.com/3021882/184988646-b065a503-2d67-4dcd-bf1e-f45e6e73beed.mov

### After - see it works

https://user-images.githubusercontent.com/3021882/185195971-2f86debd-c20a-4d07-817c-2bb53a464659.mov

### CODE REVIEW

**Comment by kmadsen** [comment]
Ran into another crash when building the "good" video in the description

```

2022-08-16 14:56:17.371 11428-12229/com.mapbox.maps.testapp.auto E/AndroidRuntime: FATAL EXCEPTION: animation_thread_20b958fc-264f-44ae-8290-890fefada0b7
    Process: com.mapbox.maps.testapp.auto, PID: 11428
    java.lang.Error: std::exception
        at com.mapbox.maps.CameraManager.setCamera(Native Method)
        at com.mapbox.maps.NativeMapImpl.setCamera(NativeMapImpl.kt:36)
        at com.mapbox.maps.MapboxMap.setCamera(MapboxMap.kt:408)
        at com.mapbox.maps.testapp.auto.car.CarCameraController.changePositionListener$lambda-1(CarCameraController.kt:34)
        at com.mapbox.maps.testapp.auto.car.CarCameraController.$r8$lambda$MSgu-k38twclsLage2OT3YN_jPw(Unknown Source:0)
        at com.mapbox.maps.testapp.auto.car.CarCameraController$$ExternalSyntheticLambda1.onIndicatorPositionChanged(Unknown Source:2)
        at com.mapbox.maps.plugin.locationcomponent.LocationComponentPluginImpl.indicatorPositionChangedListener$lambda-0(LocationComponentPluginImpl.kt:71)
        at com.mapbox.maps.plugin.locationcomponent.LocationComponentPluginImpl.$r8$lambda$LEp20bCtxkr0JjPtC5olN7XUb4M(Unknown Source:0)
        at com.mapbox.maps.plugin.locationcomponent.LocationComponentPluginImpl$$ExternalSyntheticLambda3.onIndicatorPositionChanged(Unknown Source:2)
        at com.mapbox.maps.plugin.locationcomponent.animators.PuckPositionAnimator.updateLayer(PuckPositionAnimator.kt:10)
        at com.mapbox.maps.plugin.locationcomponent.animators.PuckPositionAnimator.updateLayer(PuckPositionAnimator.kt:6)
        at com.mapbox.maps.plugin.locationcomponent.animators.PuckAnimator._init_$lambda-0(PuckAnimator.kt:34)
        at com.mapbox.maps.plugin.locationcomponent.animators.PuckAnimator.$r8$lambda$DKuRutUBCyf6yKd5UEMf-U8meWw(Unknown Source:0)
        at com.mapbox.maps.plugin.locationcomponent.animators.PuckAnimator$$ExternalSyntheticLambda0.onAnimationUpdate(Unknown Source:2)
        at android.animation.ValueAnimator.animateValue(ValueAnimator.java:1566)
        at android.animation.ValueAnimator.animateBasedOnTime(ValueAnimator.java:1357)
        at android.animation.ValueAnimator.doAnimationFrame(ValueAnimator.java:1489)
        at android.animation.AnimationHandler.doAnimationFrame(AnimationHandler.java:146)
        at android.animation.AnimationHandler.access$100(AnimationHandler.java:37)
        at android.animation.AnimationHandler$1.doFrame(AnimationHandler.java:54)
        at android.view.Choreographer$CallbackRecord.run(Choreographer.java:1140)

```

**Comment by pengdev** [comment]
> Mapbox has many calls that assume the UI thread, so in order to trigger the animations from a background thread there would be a bit work to ensure each call receives appropriate threads.

By default we require all the API surface to be called on the main thread, we enforced it in [this PR ](https://github.com/mapbox/mapbox-maps-android/pull/1316), but you can disable thread checking fully by inclusion of an Android Manifest entry:
```
        <meta-data
            android:name="com.mapbox.maps.ThreadChecker"
            android:value="false" />
```

File: plugin-animation/src/main/java/com/mapbox/maps/plugin/animation/CameraAnimationsPluginImpl.kt [hunk 1]
```diff
@@ -267,84 +274,92 @@ class CameraAnimationsPluginImpl : CameraAnimationsPlugin {
...
+          (animation as? CameraAnimator<*>)?.apply {
+            // check for a specific use-case when canceling an animation with start delay that
+            // has not yet started - in that case onAnimationStart logic must be skipped
+            if (canceled) {
+              return@post
+            }
+            lifecycleListeners.forEach {
+              it.onAnimatorStarting(type, this, owner)
+            }
+            mapTransformDelegate.setUserAnimationInProgress(true)
```
Comment (kiryldz) [inline plugin-animation/src/main/java/com/mapbox/maps/plugin/animation/CameraAnimationsPluginImpl.kt:287]: just to double check: if we create beforehand `val mainHandler = Handler(Looper.getMainLooper())` and will actually call native functions here as `mainHandler.post { mapTransformDelegate.setUserAnimationInProgress(true) }` - will this work on Xiaomi device when device is locked? If it will work - then imo it will be a bit safer to call gl-native functions from main thread while creating animators on background thread.

**Comment by kiryldz** [comment]
> Ran into another crash when building the "good" video in the description
> 
> ```
> 
> 2022-08-16 14:56:17.371 11428-12229/com.mapbox.maps.testapp.auto E/AndroidRuntime: FATAL EXCEPTION: animation_thread_20b958fc-264f-44ae-8290-890fefada0b7
>     Process: com.mapbox.maps.testapp.auto, PID: 11428
>     java.lang.Error: std::exception
>         at com.mapbox.maps.CameraManager.setCamera(Native Method)
>         at com.mapbox.maps.NativeMapImpl.setCamera(NativeMapImpl.kt:36)
>         at com.mapbox.maps.MapboxMap.setCamera(MapboxMap.kt:408)
>         at com.mapbox.maps.testapp.auto.car.CarCameraController.changePositionListener$lambda-1(CarCameraController.kt:34)
>         at com.mapbox.maps.testapp.auto.car.CarCameraController.$r8$lambda$MSgu-k38twclsLage2OT3YN_jPw(Unknown Source:0)
>         at com.mapbox.maps.testapp.auto.car.CarCameraController$$ExternalSyntheticLambda1.onIndicatorPositionChanged(Unknown Source:2)
>         at com.mapbox.maps.plugin.locationcomponent.LocationComponentPluginImpl.indicatorPositionChangedListener$lambda-0(LocationComponentPluginImpl.kt:71)
>         at com.mapbox.maps.plugin.locationcomponent.LocationComponentPluginImpl.$r8$lambda$LEp20bCtxkr0JjPtC5olN7XUb4M(Unknown Source:0)
>         at com.mapbox.maps.plugin.locationcomponent.LocationComponentPluginImpl$$ExternalSyntheticLambda3.onIndicatorPositionChanged(Unknown Source:2)
>         at com.mapbox.maps.plugin.locationcomponent.animators.PuckPositionAnimator.updateLayer(PuckPositionAnimator.kt:10)
>         at com.mapbox.maps.plugin.locationcomponent.animators.PuckPositionAnimator.updateLayer(PuckPositionAnimator.kt:6)
>         at com.mapbox.maps.plugin.locationcomponent.animators.PuckAnimator._init_$lambda-0(PuckAnimator.kt:34)
>         at com.mapbox.maps.plugin.locationcomponent.animators.PuckAnimator.$r8$lambda$DKuRutUBCyf6yKd5UEMf-U8meWw(Unknown Source:0)
>         at com.mapbox.maps.plugin.locationcomponent.animators.PuckAnimator$$ExternalSyntheticLambda0.onAnimationUpdate(Unknown Source:2)
>         at android.animation.ValueAnimator.animateValue(ValueAnimator.java:1566)
>         at android.animation.ValueAnimator.animateBasedOnTime(ValueAnimator.java:1357)
>         at android.animation.ValueAnimator.doAnimationFrame(ValueAnimator.java:1489)
>         at android.animation.AnimationHandler.doAnimationFrame(AnimationHandler.java:146)
>         at android.animation.AnimationHandler.access$100(AnimationHandler.java:37)
>         at android.animation.AnimationHandler$1.doFrame(AnimationHandler.java:54)
>         at android.view.Choreographer$CallbackRecord.run(Choreographer.java:1140)
> ```

Could be 'fixed' the same way as in https://github.com/mapbox/mapbox-maps-android/pull/652 😄

File: plugin-animation/src/main/java/com/mapbox/maps/plugin/animation/CameraAnimationsPluginImpl.kt (hunk 1 above)
Comment (kmadsen) [reply plugin-animation/src/main/java/com/mapbox/maps/plugin/animation/CameraAnimationsPluginImpl.kt]: yeah i think it works, but i'm a little unsure about the state lock `setUserAnimationInProgress`. If that is on a different thread than the rest of this function then the `nativeMap.isUserAnimationInProgress` is going to be checking on the wrong thread.

Actually, this makes makes it unnecessary to change between the main thread and the animation thread https://github.com/mapbox/mapbox-maps-android/pull/1605#issuecomment-1217939433

It seems like it would be better to designate an animation thread, and never switch to the main thread (unless you're updating a view)

**Comment by kmadsen** [comment]
@pengdev @kiryldz 

Actually, from your comments I've made a few changes that make it so this approach really could be considered for production.

The `MapboxAnimatorStarter` is by default, using the main thread. And then the `android-auto-app` specifies background thread animations in the manifest

```
        <meta-data
            android:name="com.mapbox.maps.ThreadChecker"
            android:value="false" />

        <meta-data
            android:name="com.mapbox.maps.BackgroundAnimatorThread"
            android:value="true" />
```

**Comment by kmadsen** [comment]
The `MapboxAnimatorStarter` would need to be exposed so that the downstream `NavigationCamera` could post animations to this same thread

https://github.com/mapbox/mapbox-navigation-android/blob/5ea19cd0b777b4cbe6bf63cf1409cc5ac6216ab4/libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/camera/NavigationCamera.kt#L418

File: plugin-animation/src/main/java/com/mapbox/maps/plugin/animation/CameraAnimationsPluginImpl.kt (hunk 1 above)
Comment (kiryldz) [reply plugin-animation/src/main/java/com/mapbox/maps/plugin/animation/CameraAnimationsPluginImpl.kt]: @kmadsen it's not about the thread checking. Core requires native `ALooper` be attached to the thread to allow scheduling in C++ if needed from main thread to other native threads. When you create background handler thread in Android (as done in this PR) - it does have the looper but native code does not make use of it as we're using custom run loop implementation.

As per current architecture attaching native looper is not exposed to Java (so you can't attach it to your background thread) and implemented internally for 2 threads _only_: main Android thread and our Mapbox render thread.

To sum up: e.g. `nativeMap.setCamera` does not make use of any scheduling logic (most likely but who knows 😉) under the hood but potentially some other native function invoked from background handler thread may result in a native crash. IMO it's still safer to actually run animators on background thread in Java but schedule native calls thru `mainHandler` as I described above. In that case ThreadChecker could be actually enabled.

Comment (kmadsen) [reply plugin-animation/src/main/java/com/mapbox/maps/plugin/animation/CameraAnimationsPluginImpl.kt]: > In that case ThreadChecker could be actually enabled.

There are potentially large refactors to ensure all native calls will happen on the main thread. One of them that I found was in the [CameraAnimatorsFactory](https://github.com/mapbox/mapbox-maps-android/blob/2220e35d817837a16c36a976530b57d6d7fd923f/plugin-animation/src/main/java/com/mapbox/maps/plugin/animation/CameraAnimatorsFactory.kt#L367). This part of the animation uses a [TypeEvaluator](https://developer.android.com/reference/android/animation/TypeEvaluator) which is synchronous.

If we were to force main thread there, it will also require a thread context switch in the middle of the animation. There may be a bunch of cases like this that will require careful thought before being able to enable ThreadChecker with background thread animations.

**Comment by kmadsen** [comment]
Another crash when having the `com.mapbox.maps.ThreadChecker` enabled, and background thread enabled

```
2022-08-17 16:26:06.382 28602-28683/com.mapbox.maps.testapp.auto E/AndroidRuntime: FATAL EXCEPTION: animation_thread_4a65c2a4-2aad-4f24-8ff3-9fbda277abd5
    Process: com.mapbox.maps.testapp.auto, PID: 28602
    com.mapbox.maps.exception.WorkerThreadException: The exception that is thrown when an application attempts to 
    perform a map operation on a worker thread.
        at com.mapbox.maps.ThreadChecker.throwIfNotMainThread(ThreadChecker.kt:88)
        at com.mapbox.maps.Style.checkNativeStyle(Style.kt:970)
        at com.mapbox.maps.Style.styleSourceExists(Style.kt:348)
        at com.mapbox.maps.plugin.locationcomponent.ModelSourceWrapper.updateProperty(ModelSourceWrapper.kt:60)
        at com.mapbox.maps.plugin.locationcomponent.ModelSourceWrapper.setPositionAndOrientation(ModelSourceWrapper.kt:52)
        at com.mapbox.maps.plugin.locationcomponent.ModelLayerRenderer.updateLocationOrBearing(ModelLayerRenderer.kt:96)
        at com.mapbox.maps.plugin.locationcomponent.ModelLayerRenderer.setLayerLocation(ModelLayerRenderer.kt:84)
        at com.mapbox.maps.plugin.locationcomponent.ModelLayerRenderer.setLatLng(ModelLayerRenderer.kt:64)
        at com.mapbox.maps.plugin.locationcomponent.animators.PuckPositionAnimator.updateLayer(PuckPositionAnimator.kt:9)
        at com.mapbox.maps.plugin.locationcomponent.animators.PuckPositionAnimator.updateLayer(PuckPositionAnimator.kt:6)
        at com.mapbox.maps.plugin.locationcomponent.animators.PuckAnimator._init_$lambda-0(PuckAnimator.kt:34)
        at com.mapbox.maps.plugin.locationcomponent.animators.PuckAnimator.$r8$lambda$DKuRutUBCyf6yKd5UEMf-U8meWw(Unknown Source:0)
        at com.mapbox.maps.plugin.locationcomponent.animators.PuckAnimator$$ExternalSyntheticLambda0.onAnimationUpdate(Unknown Source:2)
```

File: plugin-animation/src/main/java/com/mapbox/maps/plugin/animation/CameraAnimationsPluginImpl.kt (hunk 1 above)
Comment (kmadsen) [reply plugin-animation/src/main/java/com/mapbox/maps/plugin/animation/CameraAnimationsPluginImpl.kt:286]: https://github.com/mapbox/mapbox-maps-android/pull/1605#issuecomment-1218644974

^ Here's another example from the location animations where it crashes

File: CHANGELOG.md [hunk 2]
```diff
@@ -15,6 +15,7 @@ Mapbox welcomes participation and contributions from everyone.
 ## Features ✨ and improvements 🏁
 * Introduce a callback to be invoked when the device compass sensors need to be re-calibrated. ([1513](https://github.com/mapbox/mapbox-maps-android/pull/1513))
 * Add support for `LocationComponentSettingsInterface.pulsingMaxRadius` to follow location's accuracy radius. ([1561](https://github.com/mapbox/mapbox-maps-android/pull/1561))
+* Add experimental api to start animations from a background thread. Start animations with the `CameraAnimationsPlugin.post` function. Background thread animations is enabled with a Manifest metadata entry of `com.mapbox.maps.BackgroundAnimatorThread` and corresponding true value. This is needed for Xiaomi devices to process animations when using Android Auto. ([1605](https://github.com/mapbox/mapbox-maps-android/pull/1605)).
```
Comment (kiryldz) [inline CHANGELOG.md:18]: should be moved under `main` block above, 10.8.0-beta.1 was already released

File: android-auto-app/src/main/AndroidManifest.xml [hunk 3]
```diff
@@ -38,6 +38,14 @@
             android:value="1"
             tools:ignore="MetadataTagInsideApplicationTag" />
 
+        <meta-data
+            android:name="com.mapbox.maps.ThreadChecker"
+            android:value="false" />
+
+        <meta-data
+            android:name="com.mapbox.maps.BackgroundAnimatorThread"
+            android:value="true" />
```
Comment (kiryldz) [inline android-auto-app/src/main/AndroidManifest.xml:47]: Although background thread is used for animations - gl-native calls should still happen from the main thread. Thus this could be omitted I guess.

Another point - IMO we don't need to actually switch to background thread for animations on _all_ the devices for AA. On most devices (not Xiaomi) it works OK as it is now. It makes sense imo to hide switching animations to background thread under some API and delegate setting this flag to end users. Switching to background thread is not something safe and we should retain current behaviour (using main thread to start animations) as the default one to reduce errors.
cc @yunikkk as we were discussing that offline

File: sdk-base/src/main/java/com/mapbox/maps/plugin/animation/MapboxAnimatorThread.kt [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+}
+
+/**
+ * The [PackageManager] has issues being mocked. This class makes it so usages of the package
+ * meta data can be tested without mocking the application info metadata.
+ *
+ * https://github.com/mockk/mockk/issues/182
+ */
+@VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
+internal class AndroidManifestMetaData {
```
Comment (kiryldz) [inline sdk-base/src/main/java/com/mapbox/maps/plugin/animation/MapboxAnimatorThread.kt:109]: as described in another comment - I would not control if background thread is needed thru manifest but would make it runtime API for AA case only

File: sdk-base/src/main/java/com/mapbox/maps/plugin/animation/CameraAnimationsPlugin.kt [hunk 5]
```diff
@@ -325,4 +326,18 @@ interface CameraAnimationsPlugin : MapPlugin {
...
+   *
+   * The thread for animations can be specified in the AndroidManifest.xml.
+   * ``` xml
+   *  <meta-data
+   *    android:name="com.mapbox.maps.BackgroundAnimatorThread"
+   *    android:value="true" />
+   * ```
+   */
+  @MapboxExperimental
+  fun post(runnable: Runnable): Boolean
```
Comment (kiryldz) [inline sdk-base/src/main/java/com/mapbox/maps/plugin/animation/CameraAnimationsPlugin.kt:342]: We need background thread as error-prone solution for AA + Xiaomi case only. There's no need to expose additional method on `sdk-base` module as `extension-androidauto` has dependency on `sdk` module which has api dependencies on all other plugins. This means we could expose this method on `CameraAnimationsPluginImpl` directly.
Main idea - this solution is overall a hacky solution for minority of devices in a specific use-case (AA) and we need to make this as less visible as possible.

File: sdk-base/src/main/java/com/mapbox/maps/plugin/animation/MapboxAnimatorThread.kt [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+    return if (Looper.myLooper() == handler?.looper) {
+      runnable.run()
+      true
+    } else {
+      handler?.post(runnable) ?: false
+    }
+  }
+}
+
+private class BackgroundAnimatorThread : AnimatorThread {
```
Comment (kiryldz) [inline sdk-base/src/main/java/com/mapbox/maps/plugin/animation/MapboxAnimatorThread.kt:150]: I'd split those classes to separate files for better readability

File: sdk-base/src/main/java/com/mapbox/maps/plugin/animation/MapboxAnimatorThread.kt [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+  private val threadName = "animation_thread_${UUID.randomUUID()}"
+  private var handlerThread: HandlerThread = HandlerThread(
+    threadName,
+    Process.THREAD_PRIORITY_DISPLAY
```
Comment (kiryldz) [inline sdk-base/src/main/java/com/mapbox/maps/plugin/animation/MapboxAnimatorThread.kt:154]: why do we need such high priority? we already have render thread that uses such priority

File: sdk/src/main/java/com/mapbox/maps/MapboxMap.kt [hunk 8]
```diff
@@ -405,7 +405,17 @@ class MapboxMap :
...
+      nativeMap.setCamera(cameraOptions)
+    } catch (thr: Throwable) {
+      logW(
+        TAG,
+        """
+          setCamera failed: ${thr.message}
+          This will happen when using com.mapbox.maps.BackgroundAnimatorThread 
+        """.trimIndent()
+      )
+    }
```
Comment (kiryldz) [inline sdk/src/main/java/com/mapbox/maps/MapboxMap.kt:418]: we should not call setCamera from background thread even in AA case for Xiaomi. All animations should go thru animation plugin where this is already handled

File: plugin-animation/src/main/java/com/mapbox/maps/plugin/animation/CameraAnimationsPluginImpl.kt (hunk 1 above)
Comment (yunikkk) [reply plugin-animation/src/main/java/com/mapbox/maps/plugin/animation/CameraAnimationsPluginImpl.kt:286]: > https://github.com/mapbox/mapbox-maps-android/pull/1605#issuecomment-1218644974
> ^ Here's another example from the location animations where it crashes

to fix this one I assume we should switch back to the main thread in the update listener here
https://github.com/mapbox/mapbox-maps-android/blob/main/plugin-locationcomponent/src/main/java/com/mapbox/maps/plugin/locationcomponent/animators/PuckAnimator.kt#L33

Comment (yunikkk) [reply plugin-animation/src/main/java/com/mapbox/maps/plugin/animation/CameraAnimationsPluginImpl.kt:286]: Overal I agree with @kiryldz - we call gl-native from the main thread for a reason (and even added the thread checker in the debug builds). It seems risky and fragile to change to calling from other thread without native looper attached, but also considering possible races with the state of gl-native. 

@kmadsen would it be possible to just postMain updates in every place where `ValueAnimator.addUpdateListener` is added? E.g. `PuckAnimator` and `CameraAnimator`?

File: android-auto-app/src/main/AndroidManifest.xml (hunk 3 above)
Comment (kmadsen) [reply android-auto-app/src/main/AndroidManifest.xml:47]: Notice that this is only enabled for the `android-auto-app`. The idea is to have zero impact on the current main thread solution. The developers who are trying to release android-auto could enable the flag, to make it work.

That said, there are still a lot of issues in this PR with thread context switching. It locks, it janks, and it crashes unexpectedly. Unfortunately, the degraded experience is still better than what is happening now - it does not work at all.

Do you @yunikkk have a device that can reproduce this error? It is not a small amount of work.

File: sdk-base/src/main/java/com/mapbox/maps/plugin/animation/CameraAnimationsPlugin.kt (hunk 5 above)
Comment (kmadsen) [reply sdk-base/src/main/java/com/mapbox/maps/plugin/animation/CameraAnimationsPlugin.kt:342]: Are you suggesting that we build a Xiaomi device detector? And then behave differently with those animations?

> this solution is overall a hacky solution for minority

I actually fully agree. The goal right now, is to try and figure out how to make animations work at all. Because they don't work. 

Another solution could be, build a new animation framework for Android Auto - but that is not a small project.

File: sdk-base/src/main/java/com/mapbox/maps/plugin/animation/MapboxAnimatorThread.kt (hunk 4 above)
Comment (kmadsen) [reply sdk-base/src/main/java/com/mapbox/maps/plugin/animation/MapboxAnimatorThread.kt:109]: > runtime API for AA case only

All the issues that are exposed when this is enabled, will happen for the AA case alone. The AA case is completely messed up - so I'm using this to discover why it is so messed up.

File: plugin-animation/src/main/java/com/mapbox/maps/plugin/animation/CameraAnimationsPluginImpl.kt (hunk 1 above)
Comment (kmadsen) [reply plugin-animation/src/main/java/com/mapbox/maps/plugin/animation/CameraAnimationsPluginImpl.kt:286]: > would it be possible to just postMain updates in every place where ValueAnimator.addUpdateListener is added? E.g. PuckAnimator and CameraAnimator?

This was the reason I introduced the `postMain` function. Because, am trying to switch to the main thread everywhere. But that is not a small change, there are many places where gl-native is called. Some implementations may require a large refactor - see this comment https://github.com/mapbox/mapbox-maps-android/pull/1605#discussion_r948208432

> There are potentially large refactors to ensure all native calls will happen on the main thread. One of them that I found was in the [CameraAnimatorsFactory](https://github.com/mapbox/mapbox-maps-android/blob/2220e35d817837a16c36a976530b57d6d7fd923f/plugin-animation/src/main/java/com/mapbox/maps/plugin/animation/CameraAnimatorsFactory.kt#L367). This part of the animation uses a [TypeEvaluator](https://developer.android.com/reference/android/animation/TypeEvaluator) which is synchronous.

File: sdk/src/main/java/com/mapbox/maps/MapboxMap.kt (hunk 8 above)
Comment (kmadsen) [reply sdk/src/main/java/com/mapbox/maps/MapboxMap.kt:418]: Agree with a larger refactor or improvement. This try catch is the simplest possible solution, in order to try and make it work.

File: CHANGELOG.md (hunk 2 above)
Comment (kmadsen) [reply CHANGELOG.md:18]: Yeah, I don't think we plan on merging this anytime soon. This PR is helping us surface issues and discuss solutions.

I was adding this in case we wanted to merge `android:name="com.mapbox.maps.BackgroundAnimatorThread"`, because with it disabled there should be no change to the existing experience.

With it enabled it will fix a completely broken experience, but it still contains many issues

File: sdk-base/src/main/java/com/mapbox/maps/plugin/animation/MapboxAnimatorThread.kt (hunk 7 above)
Comment (kmadsen) [reply sdk-base/src/main/java/com/mapbox/maps/plugin/animation/MapboxAnimatorThread.kt:154]: Yeah, could play with lower priority here. Just copy-paste from `MapboxRenderThread`

There are honestly many other big issues, creating high priority is mostly to make it easier to reproduce the other issues.

File: sdk-base/src/main/java/com/mapbox/maps/plugin/animation/MapboxAnimatorThread.kt (hunk 6 above)
Comment (kmadsen) [reply sdk-base/src/main/java/com/mapbox/maps/plugin/animation/MapboxAnimatorThread.kt:150]: Yeah I was playing around with this for some time yesterday, trying to make it more readable.

What exactly is the suggestion? The classes are broken up a bit (`MapboxAnimatorThread` is using 2 other classes, at first everything was in the `MapboxAnimatorThread`)

Adding the `BackgroundAnimatorThread` made `MapboxAnimatorThread` simple
``` kotlin
  fun post(runnable: Runnable): Boolean {
    return (backgroundAnimatorThread ?: mainAnimatorThread).post(runnable)
  }
```

Comment (kmadsen) [reply sdk-base/src/main/java/com/mapbox/maps/plugin/animation/MapboxAnimatorThread.kt]: Yeah needed to move packages, moved it all to the `:plugin-animations:` module

File: sdk-base/src/main/java/com/mapbox/maps/plugin/animation/MapboxAnimatorThread.kt (hunk 7 above)
Comment (kmadsen) [reply sdk-base/src/main/java/com/mapbox/maps/plugin/animation/MapboxAnimatorThread.kt]: Going to try `THREAD_PRIORITY_FOREGROUND`

File: sdk/src/main/java/com/mapbox/maps/MapboxMap.kt (hunk 8 above)
Comment (kmadsen) [reply sdk/src/main/java/com/mapbox/maps/MapboxMap.kt:418]: Looking to ensure that `indicatorPositionChangedListener` will only be posted by the main thread, then this try/catch could be removed.

That is this crash https://github.com/mapbox/mapbox-maps-android/pull/1605#issuecomment-1217969582

Comment (kmadsen) [reply sdk/src/main/java/com/mapbox/maps/MapboxMap.kt:418]: This will crash even when it's posted from the main thread. There are more fixes needed in order to remove this try/catch. Note that this crash happens even when the ThreadChecker is false

```
2022-08-18 11:21:18.203 5923-5923/com.mapbox.maps.testapp.auto I/Mapbox: [maps-android\kyle_debug]: OnIndicatorPositionChangedListener main
2022-08-18 11:21:18.219 5923-5923/com.mapbox.maps.testapp.auto I/Mapbox: [maps-android\kyle_debug]: OnIndicatorPositionChangedListener main
2022-08-18 11:21:18.236 5923-5923/com.mapbox.maps.testapp.auto I/Mapbox: [maps-android\kyle_debug]: OnIndicatorPositionChangedListener main
2022-08-18 11:21:18.353 5923-6192/com.mapbox.maps.testapp.auto E/AndroidRuntime: FATAL EXCEPTION: animation_thread_aee6cc7a-6e7c-41a6-8790-0337587b8d24
    Process: com.mapbox.maps.testapp.auto, PID: 5923
    java.lang.Error: std::exception
        at com.mapbox.maps.CameraManager.setCamera(Native Method)
        at com.mapbox.maps.NativeMapImpl.setCamera(NativeMapImpl.kt:36)
        at com.mapbox.maps.MapboxMap.setCamera(MapboxMap.kt:408)
        at com.mapbox.maps.testapp.auto.car.CarCameraController.changeBearingListener$lambda-3(CarCameraController.kt:47)
        at com.mapbox.maps.testapp.auto.car.CarCameraController.$r8$lambda$FSsysefwAVT1AjEVdrjGVZ0YE9k(Unknown Source:0)
        at com.mapbox.maps.testapp.auto.car.CarCameraController$$ExternalSyntheticLambda0.onIndicatorBearingChanged(Unknown Source:2)
        at com.mapbox.maps.plugin.locationcomponent.LocationComponentPluginImpl.indicatorBearingChangedListener$lambda-5(LocationComponentPluginImpl.kt:146)
        at com.mapbox.maps.plugin.locationcomponent.LocationComponentPluginImpl.$r8$lambda$ENafjuhdBpO7PRaFQ5TOHHc_Y54(Unknown Source:0)
        at com.mapbox.maps.plugin.locationcomponent.LocationComponentPluginImpl$$ExternalSyntheticLambda2.onIndicatorBearingChanged(Unknown Source:2)
        at com.mapbox.maps.plugin.locationcomponent.animators.PuckBearingAnimator.updateLayer(PuckBearingAnimator.kt:12)
        at com.mapbox.maps.plugin.locationcomponent.animators.PuckBearingAnimator.updateLayer(PuckBearingAnimator.kt:5)
        at com.mapbox.maps.plugin.locationcomponent.animators.PuckAnimator._init_$lambda-0(PuckAnimator.kt:34)
        at com.mapbox.maps.plugin.locationcomponent.animators.PuckAnimator.$r8$lambda$DKuRutUBCyf6yKd5UEMf-U8meWw(Unknown Source:0)
        at com.mapbox.maps.plugin.locationcomponent.animators.PuckAnimator$$ExternalSyntheticLambda0.onAnimationUpdate(Unknown Source:2)
        at android.animation.ValueAnimator.animateValue(ValueAnimator.java:1566)
        at android.animation.ValueAnimator.animateBasedOnTime(ValueAnimator.java:1357)
        at android.animation.ValueAnimator.doAnimationFrame(ValueAnimator.java:1489)
        at android.animation.AnimationHandler.doAnimationFrame(AnimationHandler.java:146)
        at android.animation.AnimationHandler.access$100(AnimationHandler.java:37)
        at android.animation.AnimationHandler$1.doFrame(AnimationHandler.java:54)
        at android.view.Choreographer$CallbackRecord.run(Choreographer.java:1140)
        at android.view.Choreographer.doCallbacks(Choreographer.java:946)
        at android.view.Choreographer.doFrame(Choreographer.java:870)
        at android.view.Choreographer$FrameDisplayEventReceiver.run(Choreographer.java:1127)
        at android.os.Handler.handleCallback(Handler.java:938)
        at android.os.Handler.dispatchMessage(Handler.java:99)
        at android.os.Looper.loopOnce(Looper.java:210)
        at android.os.Looper.loop(Looper.java:299)
        at android.os.HandlerThread.run(HandlerThread.java:67)
```

File: sdk-base/src/main/java/com/mapbox/maps/plugin/animation/CameraAnimationsPlugin.kt (hunk 5 above)
Comment (kmadsen) [reply sdk-base/src/main/java/com/mapbox/maps/plugin/animation/CameraAnimationsPlugin.kt:342]: This is exposed so that the [NavigationCamera](https://github.com/mapbox/mapbox-navigation-android/blob/5ea19cd0b777b4cbe6bf63cf1409cc5ac6216ab4/libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/camera/NavigationCamera.kt#L418) can have access to it.

https://github.com/mapbox/mapbox-maps-android/pull/1605#issuecomment-1218166308

Are you suggesting Android Auto should build a new `NavigationCamera` so that it can run animations on a background thread? Android Auto uses the navigation-sdk, the navigation-sdk uses the maps-sdk.

File: CHANGELOG.md (hunk 2 above)
Comment (kmadsen) [reply CHANGELOG.md]: Moved it so resolving this thread

File: sdk/src/main/java/com/mapbox/maps/MapboxMap.kt (hunk 8 above)
Comment (pengdev) [reply sdk/src/main/java/com/mapbox/maps/MapboxMap.kt:418]:
```
    Process: com.mapbox.maps.testapp.auto, PID: 5923
    java.lang.Error: std::exception
        at com.mapbox.maps.CameraManager.setCamera(Native Method)
```
gl-native team is going to catch this exception on their end in the next few releases, so we could skip this crash and catch it for now.

**Comment by kmadsen** [comment]
Closing this one as it was an investigation and a new iteration will be considered which includes a runtime configuration. This approach relies on a compile time configuration, and surfaces a api for starting animations.
